File: bench/cases/allocflat/allocflat.c
#include <stdio.h>
#include <stdlib.h>

// See bench/cases/alloc/alloc.c's copy: -DBENCH_ALLOC_STATS builds a
// counting copy that prints this run's malloc count on stderr for
// bench/run.sh's cross-language comparison (#3934). The timed binary is
// built without it.
#ifdef BENCH_ALLOC_STATS
static long nmalloc = 0;
#define malloc(n) (nmalloc++, malloc(n))
#define REPORT_ALLOCS() fprintf(stderr, "[allocs] %ld\n", nmalloc)
#else
#define REPORT_ALLOCS() ((void)0)
#endif

typedef struct { long x, y, id; } Node;

int main(void) {
  int batches = 2000, per = 5000;
  long total = 0;
  for (int b = 0; b < batches; b++) {
    Node *nodes = malloc(sizeof(Node) * per);
    for (int k = 0; k < per; k++) {
      nodes[k].x = b + k;
      nodes[k].y = k + 1;
      nodes[k].id = b;
    }
    for (int k = 0; k < per; k++) total += nodes[k].x + nodes[k].y + nodes[k].id;
    free(nodes);
  }
  printf("%ld\n", total);
  REPORT_ALLOCS();
  return 0;
}

File: bench/cases/allocflat/allocflat.go
package main

import (
	"fmt"
	"os"
	"runtime"
)

type Node struct{ x, y, id int64 }

func main() {
	batches, per := 2000, 5000
	var total int64 = 0
	for b := 0; b < batches; b++ {
		nodes := make([]Node, 0, per)
		for k := 0; k < per; k++ {
			nodes = append(nodes, Node{x: int64(b + k), y: int64(k + 1), id: int64(b)})
		}
		for i := range nodes {
			total += nodes[i].x + nodes[i].y + nodes[i].id
		}
	}
	fmt.Printf("%d\n", total)
	reportAllocs()
}

// See bench/cases/alloc/alloc.go's copy: BENCH_ALLOC_STATS=1 prints this
// run's heap allocation count on stderr for bench/run.sh's cross-language
// comparison (#3934), off by default and after all timed work.
func reportAllocs() {
	if os.Getenv("BENCH_ALLOC_STATS") == "" {
		return
	}
	var ms runtime.MemStats
	runtime.ReadMemStats(&ms)
	fmt.Fprintf(os.Stderr, "[allocs] %d\n", ms.Mallocs)
}
